src/System.Private.CoreLib/shared/System/Guid.cs

```diff
@@ -682,6 +694,121 @@ private static bool TryParseExactX(ReadOnlySpan<char> guidString, ref GuidResult
             return true;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool TryParseHexD(ReadOnlySpan<char> value, out uint result)
+        {
+            Debug.Assert(value.Length == 4, "Expects exact 4 (guaranteed, because called only from TryParseExactD)");
+
+            ReadOnlySpan<byte> charToHexLookup = Number.CharToHexLookup;
+            uint numValue;
+            result = 0;
+
+            // First char
+            int num = value[0];
+            
+            // Can start only with hex digit or "+" (43 < charToHexLookup.Length); "0x" means '0' for num
+            if ((uint)num >= (uint)charToHexLookup.Length)
+                goto FalseExit;
+
+            if ((numValue = charToHexLookup[num]) == 0xFF)
+                goto FirstCharNotHex;
+
+            result = numValue;
+            
+            // Second char
+            num = value[1];
+            if ((uint)num >= (uint)charToHexLookup.Length || (numValue = charToHexLookup[num]) == 0xFF)
+            {
+                // Check for "0x", cannot use numValue for previous '0' check, so use the result
+                if ((num | 0x20) == 'x' && result == 0)
+                    goto ThirdChar;
+
+                goto FalseExit;
+            }
+                
+            result = 16 * result + numValue;
+        ThirdChar:
+            num = value[2];
+            if ((uint)num >= (uint)charToHexLookup.Length || (numValue = charToHexLookup[num]) == 0xFF)
+                goto FalseExit;
+
+            result = 16 * result + numValue;
+        FourthChar:
+            num = value[3];
+            if ((uint)num >= (uint)charToHexLookup.Length || (numValue = charToHexLookup[num]) == 0xFF)
+                goto FalseExit;
+
+            result = 16 * result + numValue;
+
+            return true;
+        FirstCharNotHex:
+            // Valid strings can be like "+123..." or "+0x1..."
+            if (num == '+')
+            {
+                num = value[1];
+                if ((uint)num >= (uint)charToHexLookup.Length || (numValue = charToHexLookup[num]) == 0xFF)
+                    goto FalseExit;
+            }
+            else
+            {
+                goto FalseExit;
+            }
+
+            // If "+0", can be "+0x"
+            if (numValue == 0)
+            {
+                num = value[2];
+                if ((num | 0x20) == 'x')
+                    goto FourthChar;
+
+                if ((uint)num >= (uint)charToHexLookup.Length || (numValue = charToHexLookup[num]) == 0xFF)
+                    goto FalseExit;
+
+                result = numValue;
+                goto FourthChar;
+            }
+
+            result = numValue;
+            goto ThirdChar;
+        FalseExit:
+            result = 0;
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Is `AggressiveInlining` helping here (does the method get inlined)? If so, it might be useful to have a second variant for parsing just two chars at a time for the `_d` ... `_k` parts to avoid unnecessary work (combining and separating the 2 bytes).

Reply:
I tried something like this before, but without noticeable performance improvement (+/- 1ns). 
Because this is counter-intuitive, I made two separate branches and run BenchmarkDotNet with three "Long" jobs for each variant (see below). Does anyone know the trick of how to get more stable results?

The same disclaimer: I have only win-x64 computer to test.

This version (3 tests results merged into the single table)

|         Method |     Mean |     Error |    StdDev |   Median |
|--------------- |---------:|----------:|----------:|---------:|
| TryParseExactD | 62.91 ns | 0.3213 ns | 1.5534 ns | 62.06 ns |
| TryParseExactD | 62.85 ns | 0.2736 ns | 1.3150 ns | 62.32 ns |
| TryParseExactD | 63.08 ns | 0.3504 ns | 1.7110 ns | 62.39 ns |
| TryParseExactN | 59.66 ns | 0.1973 ns | 0.9464 ns | 59.55 ns |
| TryParseExactN | 58.87 ns | 0.1079 ns | 0.5155 ns | 59.08 ns |
| TryParseExactN | 59.92 ns | 0.2779 ns | 1.3140 ns | 59.03 ns |

[Version with two bytes](https://github.com/OlegAxenow/coreclr/commit/f241c04cfffe45e54b9c528e1367b34d8478d9f3)

|         Method |     Mean |     Error |    StdDev |   Median |
|--------------- |---------:|----------:|----------:|---------:|
| TryParseExactD | 62.93 ns | 0.2262 ns | 1.1150 ns | 63.07 ns |
| TryParseExactD | 62.47 ns | 0.2963 ns | 1.4490 ns | 61.47 ns |
| TryParseExactD | 61.46 ns | 0.0914 ns | 0.4409 ns | 61.67 ns |
| TryParseExactN | 60.90 ns | 0.3001 ns | 1.4250 ns | 61.45 ns |
| TryParseExactN | 60.04 ns | 0.3379 ns | 1.6340 ns | 58.91 ns |
| TryParseExactN | 60.00 ns | 0.2956 ns | 1.4148 ns | 58.98 ns |

[Version with single byte](https://github.com/OlegAxenow/coreclr/commit/069604332ded9432cdb6f067528aada9f42cbd99)

|         Method |     Mean |     Error |    StdDev |   Median |
|--------------- |---------:|----------:|----------:|---------:|
| TryParseExactD | 68.15 ns | 1.4134 ns | 7.0200 ns | 63.22 ns |
| TryParseExactD | 64.17 ns | 0.2827 ns | 1.3530 ns | 63.40 ns |
| TryParseExactD | 65.26 ns | 0.2432 ns | 1.1806 ns |          |
| TryParseExactN | 60.80 ns | 0.2190 ns | 1.0690 ns | 60.33 ns |
| TryParseExactN | 60.54 ns | 0.2840 ns | 1.3570 ns | 59.87 ns |
| TryParseExactN | 59.95 ns | 0.0901 ns | 0.4331 ns |          |

Reply:
Oops, I push not completed code for TryParseExactN. I'll retest it later.

Reply:
So it looks like adding an additional overload doesn't help and current (this) version is optimal.